src/storage.rs

```diff
@@ -157,101 +392,411 @@ impl Display for StreamName {
+        // to avoid reallocations which are unnecessary.
+        for msg in this.batch.iter().skip(*this.current_index) {
+            // First, check the encoded length to avoid performing a full encode
+            // on the first message that would exceed the limit and be dropped.
+            let size = msg.encoded_len();
+            if total_size + size > MAX_BATCH_SIZE_BYTES && !serialized_rows.is_empty() {
+                break;
+            }
+
+            // Safe to encode now and include the row in this request chunk.
+            let encoded = msg.encode_to_vec();
+            debug_assert_eq!(
+                encoded.len(),
+                size,
+                "prost::encoded_len disagrees with encode_to_vec length"
+            );
+
+            serialized_rows.push(encoded);
+            total_size += size;
+            processed_count += 1;
+        }
+
+        if serialized_rows.is_empty() {
+            return Poll::Ready(None);
+        }
+
+        let proto_rows = ProtoRows { serialized_rows };
+        let proto_data = ProtoData {
+            writer_schema: if *this.include_schema_next {
+                Some(this.proto_schema.clone())
+            } else {
+                None
+            },
+            rows: Some(proto_rows),
+        };
+
+        let append_rows_request = AppendRowsRequest {
+            write_stream: this.stream_name.to_string(),
+            offset: None,
+            trace_id: this.trace_id.clone(),
+            missing_value_interpretations: HashMap::new(),
+            default_missing_value_interpretation: MissingValueInterpretation::Unspecified.into(),
+            rows: Some(append_rows_request::Rows::ProtoRows(proto_data)),
+        };
+
+        // Track the total bytes being sent using encoded_len
+        let request_bytes = append_rows_request.encoded_len();
+        this.bytes_sent_counter.fetch_add(request_bytes, Ordering::Relaxed);
+
+        *this.current_index += processed_count;
+        // After the first request, avoid sending schema again in this stream
+        if *this.include_schema_next {
+            *this.include_schema_next = false;
+        }
+
+        Poll::Ready(Some(append_rows_request))
+    }
+}
+
+/// High-level client for BigQuery Storage Write API operations.
 #[derive(Clone)]
 pub struct StorageApi {
-    write_client: BigQueryWriteClient<Channel>,
+    /// Connection pool for gRPC clients to BigQuery Storage Write API.
+    connection_pool: ConnectionPool,
+    /// Authentication provider for API requests.
     auth: Arc<dyn Authenticator>,
+    /// Base URL for BigQuery API endpoints.
     base_url: String,
 }
 
 impl StorageApi {
-    pub(crate) fn new(write_client: BigQueryWriteClient<Channel>, auth: Arc<dyn Authenticator>) -> Self {
-        Self {
-            write_client,
+    /// Creates a new storage API client instance.
+    pub(crate) async fn new(auth: Arc<dyn Authenticator>) -> Result<Self, BQError> {
```

Comment:
In the changelog, I think you should include a short summary of the breaking changes.